File: src/main/java/org/example/coin_laundry_app_backend/user/application/service/ShippingAddressService.java
package org.example.coin_laundry_app_backend.user.application.service;

import java.util.List;
import java.util.NoSuchElementException;
import lombok.RequiredArgsConstructor;
import org.example.coin_laundry_app_backend.geo.application.service.GeocodingService;
import org.example.coin_laundry_app_backend.geo.domain.model.EPSG4326Coordinate;
import org.example.coin_laundry_app_backend.user.application.record.shippingaddress.ShippingAddressSummary;
import org.example.coin_laundry_app_backend.user.domain.entity.ShippingAddress;
import org.example.coin_laundry_app_backend.user.presentation.payload.request.shippingaddress.CreateAddressRequest;
import org.example.coin_laundry_app_backend.user.presentation.payload.request.shippingaddress.UpdateAddressRequest;
import org.example.coin_laundry_app_backend.user.repository.ShippingAddressRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import reactor.core.publisher.Mono;

@Service
@Transactional
@RequiredArgsConstructor
public class ShippingAddressService {

    private final ShippingAddressRepository shippingAddressRepository;
    private final GeocodingService geocodingService;

    public Mono<List<ShippingAddressSummary>> getAllShippingAddresses(Long userId) {
        return shippingAddressRepository.findByUserId(userId)
            .map(ShippingAddressSummary::of)
            .collectList();
    }

    public Mono<ShippingAddress> getShippingAddressById(Long userId, Long addressId) {
        return shippingAddressRepository.findByIdAndUserId(addressId, userId);
    }

    public Mono<ShippingAddress> getDefaultShippingAddress(Long userId) {
        return shippingAddressRepository.findByUserIdAndIsDefaultAddressTrue(userId);
    }

    public Mono<ShippingAddress> addShippingAddress(Long userId, CreateAddressRequest request) {
        return shippingAddressRepository.countByUserId(userId)
            .flatMap(count -> {
                ShippingAddress shippingAddress = request.toEntity(userId);
                if (count == 0) {
                    shippingAddress.markAsDefaultAddress();
                }
                return geocodingService.getGeocoding(shippingAddress.getBaseAddress())
                    .switchIfEmpty(Mono.error(new NoSuchElementException("Invalid Address")))
                    .map(geoModels -> geoModels.get(0))
                    .map(geoModel -> {
                        EPSG4326Coordinate coordinates = geoModel.coordinate();
                        shippingAddress.updateCoordinates(coordinates.latitude(),
                            coordinates.longitude());
                        return shippingAddress;
                    });
            })
            .flatMap(shippingAddressRepository::save);
    }

    public Mono<ShippingAddress> updateShippingAddress(Long userId, Long addressId,
        UpdateAddressRequest request) {
        return shippingAddressRepository.findByIdAndUserId(addressId, userId)
            .flatMap(existingAddress -> {
                String previousBaseAddress = existingAddress.getBaseAddress();
                existingAddress.overwrite(request.addressLabel(),
                    request.recipientName(),
                    request.recipientPhone(),
                    request.baseAddress(),
                    request.detailAddress(),
                    request.deliveryNotes(),
                    request.entranceType(),
                    request.entranceDetail());
                if (!previousBaseAddress.equals(request.baseAddress())) {
                    return geocodingService.getGeocoding(existingAddress.getBaseAddress())
                        .switchIfEmpty(Mono.error(new NoSuchElementException("Invalid Address")))
                        .map(geoModels -> geoModels.get(0))
                        .map(geoModel -> {
                            EPSG4326Coordinate coordinates = geoModel.coordinate();
                            existingAddress.updateCoordinates(coordinates.latitude(),
                                coordinates.longitude());
                            return existingAddress;
                        })
                        .flatMap(shippingAddressRepository::save);
                }
                return shippingAddressRepository.save(existingAddress);
            });
    }

    public Mono<Void> deleteShippingAddress(Long userId, Long addressId) {
        return shippingAddressRepository.findByIdAndUserId(addressId, userId)
            .flatMap(shippingAddress -> {
                if (Boolean.TRUE.equals(shippingAddress.getIsDefaultAddress())) {
                    return Mono.error(
                        new RuntimeException("default shipping-address cannot be deleted"));
                }
                return Mono.just(shippingAddress);
            })
            .flatMap(shippingAddressRepository::delete);
    }


    public Mono<Void> setDefaultShippingAddress(Long userId, Long newDefaultAddressId) {
        return shippingAddressRepository.updateShippingAddressToDefault(userId,
            newDefaultAddressId);
    }
}

File: carry-user/src/main/kotlin/com/carry/user/application/service/ShippingAddressService.kt
package com.carry.user.application.service

import com.carry.common.exception.BusinessException
import com.carry.common.exception.ErrorCode
import com.carry.user.application.dto.CreateShippingAddressCommand
import com.carry.user.application.dto.ShippingAddressResponse
import com.carry.user.application.dto.UpdateShippingAddressCommand
import com.carry.user.application.port.inbound.ShippingAddressUseCase
import com.carry.user.domain.model.ShippingAddress
import com.carry.user.domain.repository.ShippingAddressRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class ShippingAddressService(
    private val shippingAddressRepository: ShippingAddressRepository
) : ShippingAddressUseCase {

    override fun getAddresses(userId: Long): List<ShippingAddressResponse> {
        return shippingAddressRepository.findByUserId(userId)
            .map { ShippingAddressResponse.from(it) }
    }

    @Transactional
    override fun createAddress(userId: Long, command: CreateShippingAddressCommand): ShippingAddressResponse {
        val isFirst = shippingAddressRepository.countByUserId(userId) == 0L
        val address = ShippingAddress(
            userId = userId,
            alias = command.alias,
            roadAddress = command.roadAddress,
            detailAddress = command.detailAddress,
            zipCode = command.zipCode,
            latitude = command.latitude,
            longitude = command.longitude,
            isDefault = isFirst
        )
        return ShippingAddressResponse.from(shippingAddressRepository.save(address))
    }

    @Transactional
    override fun updateAddress(userId: Long, addressId: Long, command: UpdateShippingAddressCommand): ShippingAddressResponse {
        val address = findUserAddress(userId, addressId)
        address.update(command.alias, command.roadAddress, command.detailAddress, command.zipCode, command.latitude, command.longitude)
        return ShippingAddressResponse.from(address)
    }

    @Transactional
    override fun deleteAddress(userId: Long, addressId: Long) {
        val address = findUserAddress(userId, addressId)
        shippingAddressRepository.delete(address)
    }

    @Transactional
    override fun setDefaultAddress(userId: Long, addressId: Long) {
        val current = shippingAddressRepository.findByUserIdAndIsDefaultTrue(userId)
        current?.unsetDefault()
        val address = findUserAddress(userId, addressId)
        address.setAsDefault()
    }

    private fun findUserAddress(userId: Long, addressId: Long): ShippingAddress {
        val address = shippingAddressRepository.findById(addressId)
            .orElseThrow { BusinessException(ErrorCode.NOT_FOUND, "Shipping address not found") }
        if (address.userId != userId) {
            throw BusinessException(ErrorCode.FORBIDDEN, "Not your address")
        }
        return address
    }
}
